### PR TITLE
Add `str_increment` and `str_decrement` functions

### LINKED ISSUE #452: Missing str_increment and str_decrement polyfill
These functions have been added to PHP 8.3

Help Wanted!

### PR DESCRIPTION
**Pull request by Fan2Shrek**
Add implementations of `str_increment` and `str_decrement` functions

PS: It's my first PR to open source, feel free to give me feedback !

Fixes https://github.com/symfony/polyfill/issues/452

### CODE REVIEW

File: src/Php83/Php83.php [hunk 1]
```diff
@@ -12,7 +12,7 @@
 namespace Symfony\Polyfill\Php83;
 
 /**
- * @author Ion Bazan <ion.bazan@gmail.com>
+ * @author Ion Bazan <ion.bazan@gmail.com>, Pierre Ambroise <pierre27.ambroise@gmail.com>
```
Comment (94noni) [inline src/Php83/Php83.php]: usually on a new line :)

File: tests/Php83/Php83Test.php [hunk 2]
```diff
@@ -46,7 +46,7 @@ public function testMbStrPad(string $expectedResult, string $string, int $length
     public function testMbStrPadInvalidArguments(string $expectedError, string $string, int $length, string $padString, int $padType, string $encoding = null): void
     {
         $this->expectException(\ValueError::class);
-        $this->expectErrorMessage($expectedError);
+        $this->expectExceptionMessage($expectedError);
```
Comment (94noni) [inline tests/Php83/Php83Test.php]: is this related to the PR?

File: tests/Php83/Php83Test.php [hunk 3]
```diff
@@ -202,4 +202,66 @@ public function testDateTimeExceptionClassesExist()
         $this->assertTrue(class_exists(\DateMalformedIntervalStringException::class));
         $this->assertTrue(class_exists(\DateMalformedPeriodStringException::class));
     }
+
+    /** 
+     * @dataProvider strIncrementProvider
```
Comment (94noni) [inline tests/Php83/Php83Test.php:209]: perhaps add also the `@covers` ?

File: src/Php83/Php83.php [hunk 4]
```diff
@@ -82,4 +83,66 @@ public static function mb_str_pad(string $string, int $length, string $pad_strin
...
+    {
+        if ('' === $string) {
+            throw new \ValueError('str_decrement(): Argument #1 ($string) cannot be empty');
+        }
+
+        if (1 === preg_match('/[^\x00-\x7F]/', $string)) {
+            throw new \ValueError('str_decrement(): Argument #1 ($string) must be composed only of alphanumeric ASCII characters');
+        }
+        
+        if (\in_array($string, ['A', 'a', 0], true)) {
```
Comment (stof) [inline src/Php83/Php83.php]: shoudln't this be `'0'` instead of `0` ? A string won't ever be strictly equal to an integer

File: tests/Php83/Php83Test.php [hunk 5]
```diff
@@ -202,4 +202,74 @@ public function testDateTimeExceptionClassesExist()
...
+    {
+        $this->assertSame($result, str_decrement($string));
+    }
+    
+    public static function strIncrementProvider(): iterable
+    {
+        yield ['ABD', 'ABC'];
+        yield ['EB', 'EA'];
+        yield ['AAA', 'ZZ'];
+        yield ['AAA2', 'AAA1'];
```
Comment (stof) [inline tests/Php83/Php83Test.php]: I suggest adding more tests in those data providers, taking all the cases covered by tests in https://github.com/php/php-src/pull/10358

See https://github.com/php/php-src/pull/10358/files#diff-a892b2d0a9411e1ad34917b112f49c90d4e33f3d61a8491293178d3a5a8fc53b for the tests

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Php83/Php83.php [hunk 6]
```diff
@@ -82,4 +83,116 @@ public static function mb_str_pad(string $string, int $length, string $pad_strin
...
+    {
+        if ('' === $string) {
+            throw new \ValueError('str_decrement(): Argument #1 ($string) cannot be empty');
+        }
+
+        if (!\ctype_alnum($string)) {
+            throw new \ValueError('str_decrement(): Argument #1 ($string) must be composed only of alphanumeric ASCII characters');
+        }
+
+        if (\preg_match('/^0[aA0]?$|^[aA]$/', $string)) {
```
Comment (zonuexe) [inline src/Php83/Php83.php]:
```suggestion
        if (\preg_match('/\A(?:0[aA0]?|[aA])\z/', $string)) {
```

[PHP Manual](https://www.php.net/manual/en/regexp.reference.meta.php) says:

> | Meta-character | Description |
> |----------------|-------------|
> | `$` | assert end of subject or before a terminating newline (or end of line, in multiline mode) |

It is safer to use the pair `\A` and `\z` instead of `^` and `$` in [PCRE regular expressions](https://www.php.net/manual/en/regexp.reference.escape.php).

Comment (Fan2Shrek) [reply src/Php83/Php83.php]: Done thanks for the tips 👍

File: src/Php83/Php83.php [hunk 7]
```diff
@@ -82,4 +83,116 @@ public static function mb_str_pad(string $string, int $length, string $pad_strin
...
+        return ++$string;
+    }
+
+    public static function str_decrement(string $string): string
+    {
+        if ('' === $string) {
+            throw new \ValueError('str_decrement(): Argument #1 ($string) cannot be empty');
+        }
+
+        if (!\ctype_alnum($string)) {
```
Comment (nicolas-grekas) [inline src/Php83/Php83.php]: it'd be great to get rid of any calls to ctype so that the code doesn't rely 
preg_match on `^[a-zA-Z0-9]+$` should do it (same below)

File: src/Php83/Php83.php [hunk 8]
```diff
@@ -82,4 +83,116 @@ public static function mb_str_pad(string $string, int $length, string $pad_strin
...
+
+        if (\preg_match('/\A(?:0[aA0]?|[aA])\z/', $string)) {
+            throw new \ValueError(sprintf('str_decrement(): Argument #1 ($string) "%s" is out of decrement range', $string));
+        }
+
+        if (!\in_array(substr($string, -1), ['A', 'a', '0'], true)) {
+            return join('', array_slice(str_split($string), 0, -1)) . chr(ord(substr($string, -1)) - 1);
+        }
+
+        $toConvert = strrev($string);
```
Comment (nicolas-grekas) [inline src/Php83/Php83.php]: it'd be great for performance and memory to get rid of the double strrev, doable?

Comment (Fan2Shrek) [reply src/Php83/Php83.php]: done :)

**Comment by nicolas-grekas** [comment]
Thank you @Fan2Shrek.
